### PR TITLE
fix(models): compose DFT-D3 virial/forces in pipeline autograd groups

### LINKED ISSUE #165: 🐛[BUG]: DFT-D3 virial dropped in PipelineModelWrapper autograd groups (composed stress omits dispersion)
### Version

0.2.0 (also reproduces on latest `main`)

### On which installation method(s) does this occur?

Source

### Describe the issue

`DFTD3ModelWrapper` computes an analytic kernel virial, but when it is a step in a `PipelineModelWrapper` autograd group (`PipelineGroup(use_autograd=True)`) that virial is silently dropped, so the composite stress omits the entire dispersion contribution. In a dense liquid this flips the sign of the pressure and breaks NPT / variable-cell for any `base + D3` composition. Forces are affected by the same mechanism, but D3's force contribution is small so it goes unnoticed.

On a 192-atom periodic water box, D3's standalone virial is **−7558 bar** and a numerical `−dE/dV` of the D3 energy agrees (**−7561 bar**), but the same D3 term inside an autograd group contributes **≈ 0** — i.e. its virial is missing. **Expected:** the composed stress includes D3's dispersion virial (≈ the standalone value).

**Root cause.** The pipeline builds each derivative in an autograd group from two summed sources: (a) autograd of the summed energy vs the affine strain, and (b) the analytic outputs each sub-model reports via `direct_derivative_keys()`.

1. In an autograd group, `PipelineModelWrapper._configure_sub_models` strips `{"forces","stress"}` from every sub-model except its `direct_derivative_keys()` (`nvalchemi/models/pipeline.py:702`).
2. `DFTD3ModelWrapper` does not override `direct_derivative_keys()`, so it inherits the `BaseModelMixin` default, which returns an empty set (`nvalchemi/models/base.py:501`). `PMEModelWrapper` does override it (`nvalchemi/models/pme.py:490`). D3's analytic virial — computed at `nvalchemi/models/dftd3.py:743` (`compute_virial = "stress" in active_outputs`) — is therefore stripped and never summed via source (b).
3. Source (a) cannot recover it: `apply_strain` scales positions and the cell but not `neighbor_matrix_shifts` (`nvalchemi/models/_utils.py:237-238`), and those shifts are absolute Cartesian vectors `n @ cell` built on the unstrained cell (`nvalchemi/data/atomic_data.py:219`); D3's kernel energy also carries no autograd graph. So `∂E_D3/∂strain ≈ 0`.

Net: D3 contributes ~0 to both sources → the dispersion virial vanishes from the composite stress.

**Proposed fix.** Add a `direct_derivative_keys()` override to `DFTD3ModelWrapper`, mirroring `PMEModelWrapper`, so its analytic forces/virial are kept and summed rather than stripped. No double counting, because D3's energy is detached, so the group autograd contributes nothing for D3:

```python
def direct_derivative_keys(self) -> set[str]:
    keys: set[str] = set()
    if "forces" in self.model_config.outputs:
        keys.add("forces")
    if "stress" in self.model_config.outputs:
        keys.add("stress")
    return keys
```

### Minimum reproducible example

Needs only `nvalchemi-toolkit` + `nvalchemi-toolkit-ops` (no model weights). It gives a structural proof — D3 reports no analytic derivatives, so the autograd group strips its `stress` — and a numerical proof — D3's virial (−7558 bar standalone) drops to ≈ 0 inside a group.

```python
import torch
from ase import Atoms
from ase.build import molecule

from nvalchemi.data import AtomicData, Batch
from nvalchemi.dynamics import NVE
from nvalchemi.dynamics.base import DynamicsStage
from nvalchemi.hooks import NeighborListHook
from nvalchemi.models import PipelineModelWrapper
from nvalchemi.models.demo import DemoModel, DemoModelWrapper
from nvalchemi.models.dftd3 import DFTD3ModelWrapper
from nvalchemi.models.pipeline import PipelineGroup, PipelineStep


def d3():
    m = DFTD3ModelWrapper(a1=0.566, a2=3.128, s8=0.3908, s6=1.0, cutoff=6.0).cuda()
    m.model_config.active_outputs = {"energy", "forces", "stress"}
    return m


def group(*steps):
    return PipelineModelWrapper(groups=[PipelineGroup(steps=list(steps), use_autograd=True)])


# 1) Structural proof: D3 reports no analytic derivatives, so the autograd group strips its stress.
step = PipelineStep(d3())
pipe = group(step)
print("D3.direct_derivative_keys() :", sorted(d3().direct_derivative_keys()), " (expected ['forces', 'stress'])")
print("D3 active outputs in group  :", sorted(pipe._step_active_overrides[id(step)]), " ('stress' stripped)")

# 2) Numerical proof: the composed stress omits D3's virial.
water = molecule("H2O")
box = Atoms()
for i in range(4):
    for j in range(4):
        for k in range(4):
            mol = water.copy()
            mol.translate([3.1 * i, 3.1 * j, 3.1 * k])
            box += mol
box.set_cell([12.4, 12.4, 12.4])
box.set_pbc(True)


def pressure(model):
    data = AtomicData.from_atoms(box, device="cuda", dtype=torch.float32)
    n = data.num_nodes
    data.forces = torch.zeros(n, 3, device="cuda")
    data.energy = torch.zeros(1, 1, device="cuda")
    data.stress = torch.zeros(1, 3, 3, device="cuda")
    batch = Batch.from_data_list([data])
    # NVE is only a harness here: it runs the neighbor-list hook + one compute(); no MD is done.
    integ = NVE(model=model, dt=1.0, n_steps=1)
    hooks = (model.make_neighbor_hooks() if hasattr(model, "make_neighbor_hooks")
             else [NeighborListHook(model.model_config.neighbor_config, stage=DynamicsStage.BEFORE_COMPUTE)])
    for h in hooks:
        integ.register_hook(h)
    integ._call_hooks(DynamicsStage.BEFORE_COMPUTE, batch)
    integ.compute(batch)
    return -float(torch.diagonal(batch.stress.reshape(3, 3)).mean()) / 6.241509e-7  # bar


base = DemoModelWrapper(DemoModel()).cuda().eval()  # any autograd-energy model, only to seed the graph
p_ref = pressure(d3())
p_in_group = pressure(group(PipelineStep(base), PipelineStep(d3()))) - pressure(group(PipelineStep(base)))
print("D3 standalone virial        : %+.1f bar" % p_ref)
print("D3 contribution in a group  : %+.1f bar  (should equal standalone; ~0 = BUG)" % p_in_group)
```

### Relevant log output

```shell
D3.direct_derivative_keys() : []  (expected ['forces', 'stress'])
D3 active outputs in group  : ['energy']  ('stress' stripped)
D3 standalone virial        : -7558.3 bar
D3 contribution in a group  : -1.0 bar  (should equal standalone; ~0 = BUG)
```

### Environment details

```shell
+ Environment location: Bare-metal
+ GPU: NVIDIA A10
+ CUDA version: 13.0 (torch cu130 build)
+ Python version: 3.12.3
+ PyTorch version: 2.12.0+cu130
+ NVIDIA Warp version: 1.14.0 (warp-lang)
+ nvalchemi-toolkit: 0.2.0
+ nvalchemi-toolkit-ops: 0.4.1
```

### PR DESCRIPTION
**Pull request by FabioFalcioni**
# ALCHEMI Toolkit Pull Request

## Description

Add `DFTD3ModelWrapper.direct_derivative_keys()` so the DFT-D3 wrapper reports the forces and virial it computes analytically in its Warp kernel. In a `PipelineModelWrapper` autograd group, `_configure_sub_models` strips `{"forces", "stress"}` from every sub-model except the keys returned by `direct_derivative_keys()`, then sums the kept analytic outputs with the group's autograd derivatives. The D3 wrapper previously inherited the `BaseModelMixin` default (an empty set), so its analytic virial was stripped and never summed, and any `base + D3` composition reported a stress that omitted the dispersion contribution. This mirrors `PMEModelWrapper.direct_derivative_keys()`. There is no double counting because the D3 kernel energy carries no autograd graph, so the group autograd contributes nothing for D3.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #165

## Changes Made

- Add `DFTD3ModelWrapper.direct_derivative_keys()` returning `{"forces", "stress"}` intersected with the declared outputs, mirroring `PMEModelWrapper`.
- Keep D3's analytic kernel forces and virial in a `PipelineModelWrapper` autograd group so the composed stress includes the dispersion contribution instead of dropping it.

## Testing

Ran `pytest test/models/test_dftd3.py` locally (all passing). `ruff check`, `ruff format --check`, and `black --check` are clean on the changed file.

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [ ] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

None.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Comment by laserkelvin** [comment]
/ok to test [974d4c5](https://github.com/NVIDIA/nvalchemi-toolkit/pull/166/commits/974d4c54cbd042343a19f675442f1047fa4bb3e3)

**Review by dallasfoster — APPROVED** [review]
So the LennardJonesModelWrapper also needs the same fix. I'll approve this PR but if you would be so kind as to make the appropriate fix to that model here or in another PR I would be grateful.

**Comment by FabioFalcioni** [comment]
Added LJ fix as well :)

**Comment by dallasfoster** [comment]
In `nvalchemi/models/lj.py` — LJ's op returns tensors (unlike D3's `mutates_args=(...) -> None`), so its returns land in the graph and backward has no formula; detach them at both call sites:

```python
# after line 459, the lj_energy_forces_virial_batch(...) call
atomic_energies, forces, virial = (
    atomic_energies.detach(), forces.detach(), virial.detach()
)
```
and
```python
# after line 479, the lj_energy_forces_batch(...) call
atomic_energies, forces = atomic_energies.detach(), forces.detach()
```
